projects/concurrency/main.go

```diff
@@ -0,0 +1,111 @@
+package main
+
+import "fmt"
+
+type Value[K comparable] struct {
+	data any
+	next *K
+	prev *K
+}
+
+type Cache[K comparable] struct {
+	size  int
+	m     map[K]Value[K]
+	first *K
+	last  *K
+}
+
+func NewCache[K comparable](entryLimit int) Cache[K] { //All K should be unique && same type && comparable type like primitive that can compare with each other
+	return Cache[K]{
+		size:  entryLimit,
+		m:     make(map[K]Value[K]),
+		first: nil,
```

Comment:
I think your code would probably be simpler if you had two fields here, a `map` and a `LinkedList` (where you may need to implement `LinkedList` yourself, or may be able to use a library).

Right now, you have a linked list implementation in-line in your cache. While this can work, it has three big drawbacks:
1. It's harder to understand the code. Seeing `c.list.PushBack(value)` is easier to understand than updating `last` and also maybe updating `first` if the list is empty and maybe also updating `next` for some existing value and maybe also updating `prev` for some existing value. _Encapsulating_ the linked list in its own type means we can _abstract_ over it, and think in higher level terms. It's also easier to test the linked list on its own.
2. It's harder to re-use the linked list. If you've implemented a good linked list, you may want to use it for other projects. That's really easy to do if it's a standalone type, and really hard to do if it's mixed in with your cache implementation.
3. It's harder to use someone else's linked list implementation, or swap it with some other data structure. For instance, the Go standard library actually contains a linked list implementation (https://pkg.go.dev/container/list) which you could use if you wanted. If you write your own linked list separately, it's easy to replace just that one part with someone else's implementation. But it's going to be harder to disentangle the current implementation.